price-estimator/src/filter.rs

```diff
@@ -0,0 +1,119 @@
+use std::num::ParseIntError;
+use warp::Filter;
+
+/// Validate a request of the form
+/// `/markets/<baseTokenId>-<quoteTokenId>/estimated-buy-amount/<sellAmountInQuoteToken>`
+/// and extract the url parameters.
+pub fn estimated_buy_amount(
+) -> impl Filter<Extract = (TokenPair, f64), Error = warp::Rejection> + Copy {
+    warp::path!("markets" / TokenPair / "estimated-buy-amount" / f64)
+}
+
+#[derive(Debug, Copy, Clone)]
+pub struct TokenPair {
+    buy_token_id: u16,
+    sell_token_id: u16,
+}
+
+impl std::convert::Into<pricegraph::TokenPair> for TokenPair {
+    fn into(self) -> pricegraph::TokenPair {
+        pricegraph::TokenPair {
+            buy: self.buy_token_id,
+            sell: self.sell_token_id,
+        }
+    }
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum ParseTokenPairError {
+    #[error("wrong number of tokens")]
+    WrongNumberOfTokens,
+    #[error("parse int error")]
+    ParseIntError(#[from] ParseIntError),
+}
+
+impl std::str::FromStr for TokenPair {
+    type Err = ParseTokenPairError;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut split = s.split('-');
+        let mut next_token_id = || -> Result<u16, ParseTokenPairError> {
+            let token_string = split
+                .next()
+                .ok_or(ParseTokenPairError::WrongNumberOfTokens)?;
+            token_string.parse().map_err(From::from)
+        };
+        let buy_token_id = next_token_id()?;
+        let sell_token_id = next_token_id()?;
+        if next_token_id().is_ok() {
```

Comment:
I don't think this is enough as `1-2-abc` will be accepted.

Reply:
How important is it to reject all wrong queries? I feel we shouldn't do more work to reject them but I also think that in general it is better to be more stingy with the api.

Reply:
Personally, I think its better to be more restrictive with accepted input then less restrictive. It also helps avoid things like breaking changes.

Reply:
Should be fixed now.

Reply:
Just curious, any reason why you didn't want to use regular expressions? I feel like they would have been a very nice fit here.

Reply:
Felt like overkill to depend on regex for finding two numbers separated by `-`.